File: Visual Studio Projects/CreateColorsAndShapes/CreateColorsAndShapesCS/CreateColorsAndShapes.cs
/* -----------------------------------------------------------------------
 * <copyright file="Main.cs" company="Hybrid Software Helix Ltd">
 *  Copyright (C) 2025 Hybrid Software Helix Ltd. All rights reserved.
 * </copyright>
 * <summary>
 *  This example is provided on an "as is" basis and without warranty of any kind.
 *  Hybrid Software Helix Ltd. does not warrant or make any representations
 *  regarding the use or results of use of this example.
 * </summary>
 * -----------------------------------------------------------------------
 */

using JawsMako;

namespace CreateShapesCS;

class Program
{
    static int Main(string[] args)
    {
        if (args.Length != 1)
        {
            Console.Error.WriteLine($"Usage: {AppDomain.CurrentDomain.FriendlyName} <icc-profile>");
            return 1;
        }

        try
        {
            var iccProfilePath = args[0];
            if (!File.Exists(iccProfilePath))
                throw new FileNotFoundException("ICC profile was not found.", iccProfilePath);

            using var mako = IJawsMako.create();
            IJawsMako.enableAllFeatures(mako);

            using var assembly = IDocumentAssembly.create(mako);
            using var document = IDocument.create(mako);
            assembly.appendDocument(document);
            using var page = IPage.create(mako);
            document.appendPage(page);
            using var fixedPage = IDOMFixedPage.create(mako);
            page.setContent(fixedPage);

            // Create some process color brushes
            using var cyan = IDOMSolidColorBrush.createSolidCmyk(mako, 1.0f, 0.0f, 0.0f, 0.0f);
            using var magenta = IDOMSolidColorBrush.createSolidCmyk(mako, 0.0f, 1.0f, 0.0f, 0.0f);
            using var yellow = IDOMSolidColorBrush.createSolidCmyk(mako, 0.0f, 0.0f, 1.0f, 0.0f);
            using var black = IDOMSolidColorBrush.createSolidCmyk(mako, 0.0f, 0.0f, 0.0f, 1.0f);

            // Create some spot color brushes with types LAB, ICCBased and DeviceN

            // Create an LAB colorspace with D65 white point and -128 to 127 ranges for a and b values
            using var labColorSpace = IDOMColorSpaceLAB.create(mako, 0.9504f, 1.0f, 1.0888f, 0.0f, 0.0f, 0.0f, -128, 127, -128, 127);

            // Create a device CMYK colorspace from an ICC profile
            using var iccBasedColorSpace = IDOMColorSpaceICCBased.create(
                mako, IDOMICCProfile.create(
                    mako, IInputStream.createFromFile(
                        mako, iccProfilePath)));

            // Create a LAB color
            using var pantoneBlue072C_lab = IDOMColor.create(mako, labColorSpace, 1.0, 17.64, 43.0, -76.0);

            // Make a copy and convert to ICC space
            using var pantoneBlue072C_fogra = IDOMColor.fromRCObject(pantoneBlue072C_lab.clone(mako));
            pantoneBlue072C_fogra.setColorSpace(iccBasedColorSpace, eRenderingIntent.eRelativeColorimetric, eBlackPointCompensation.eBPCDefault, mako);

            // Create a DeviceN color
            using var pantoneBlue072C_spot = MakeDeviceNColor(mako, "PANTONE BLUE 072 C", new CEDLVectDouble(new double[] { 17.64, 43.0, -76.0 }), labColorSpace);

            // Create an All spot color
            using var allColor = MakeDeviceNColor(mako, "All", new CEDLVectDouble(new[] { 1.0, 1.0, 1.0, 1.0 }), IDOMColorSpaceDeviceCMYK.create(mako));

            // Create spot color brushes
            using var labBrush = IDOMSolidColorBrush.create(mako, pantoneBlue072C_lab);
            using var iccBrush = IDOMSolidColorBrush.create(mako, pantoneBlue072C_fogra);
            using var deviceNBrush = IDOMSolidColorBrush.create(mako, pantoneBlue072C_spot);
            using var allBrush = IDOMSolidColorBrush.create(mako, allColor);

            // Set box size and origin
            using var masterBoxSize = new FPoint(90, 90);
            using var boxSize = new FPoint(masterBoxSize);
            using var origin = new FPoint(85, 85);
            using var start = new FPoint(origin);

            // Draw Cyan box
            var box = new FRect(start.x, start.y, boxSize.x, boxSize.y);
            fixedPage.appendChild(IDOMPathNode.createFilled(mako, IDOMPathGeometry.create(mako, box), cyan));
            start.x += boxSize.x;

            // Draw Magenta box
            box = new FRect(start.x, start.y, boxSize.x, boxSize.y);
            fixedPage.appendChild(IDOMPathNode.createFilled(mako, IDOMPathGeometry.create(mako, box), magenta));
            start.x += boxSize.x;

            // Draw Yellow box
            box = new FRect(start.x, start.y, boxSize.x, boxSize.y);
            fixedPage.appendChild(IDOMPathNode.createFilled(mako, IDOMPathGeometry.create(mako, box), yellow));
            start.x += boxSize.x;

            // Draw Black box
            box = new FRect(start.x, start.y, boxSize.x, boxSize.y);
            fixedPage.appendChild(IDOMPathNode.createFilled(mako, IDOMPathGeometry.create(mako, box), black));
            start.x += boxSize.x;

            // Draw LAB box
            box = new FRect(start.x, start.y, boxSize.x, boxSize.y);
            fixedPage.appendChild(IDOMPathNode.createFilled(mako, IDOMPathGeometry.create(mako, box), labBrush));
            start.x += boxSize.x;

            // Draw ICC box
            box = new FRect(start.x, start.y, boxSize.x, boxSize.y);
            fixedPage.appendChild(IDOMPathNode.createFilled(mako, IDOMPathGeometry.create(mako, box), iccBrush));
            start.x += boxSize.x;

            // Draw DeviceN box
            box = new FRect(start.x, start.y, boxSize.x, boxSize.y);
            fixedPage.appendChild(IDOMPathNode.createFilled(mako, IDOMPathGeometry.create(mako, box), deviceNBrush));
            start.x += boxSize.x;

            // Draw a border in All
            const uint strokeWidth = 20;
            box = new FRect(origin.x - strokeWidth / 2.0, origin.y - strokeWidth / 2.0, (boxSize.x * 7) + strokeWidth, boxSize.y + strokeWidth);
            fixedPage.appendChild(IDOMPathNode.createStroked(mako, IDOMPathGeometry.create(mako, box), allBrush, new FMatrix(),
                IDOMPathGeometry.Null(), strokeWidth));

            // Draw some other shapes below
            origin.y += 180;

            DrawRow(fixedPage, mako, origin, masterBoxSize, [cyan, magenta, yellow, black], ShapeType.Ellipse);
            origin.y += 200;
            DrawRow(fixedPage, mako, origin, masterBoxSize, [labBrush, iccBrush, deviceNBrush, allBrush], ShapeType.Hexagon);
            origin.y += 200;
            DrawRow(fixedPage, mako, origin, masterBoxSize, [cyan, magenta, yellow, black], ShapeType.Polygon, 8, 22.5);
            origin.y += 200;
            DrawRow(fixedPage, mako, origin, masterBoxSize, [labBrush, iccBrush, deviceNBrush, allBrush], ShapeType.Target, 0, 0.0, 3);

            using var pdf = IPDFOutput.create(mako);
            pdf.writeAssembly(assembly, "test.pdf");
        }
        catch (MakoException e)
        {
            Console.WriteLine($"MakoException: {e.m_msg}");
            return 1;
        }
        catch (Exception e)
        {
            Console.WriteLine($"Exception: {e.Message}");
            return 1;
        }

        return 0;
    }

    private enum ShapeType { Ellipse, Hexagon, Polygon, Target }

    private static void DrawRow(IDOMFixedPage page, IJawsMako mako, FPoint origin, FPoint masterBoxSize,
        IDOMBrush[] brushes,
        ShapeType type, int sides = 0, double angle = 0, int lines = 0)
    {
        var boxSize = new FPoint(masterBoxSize.x, masterBoxSize.y);
        var start = new FPoint(origin.x, origin.y);

        foreach (var brush in brushes)
        {
            // Draw stroked version first
            var strokeBox = new FRect(start.x, start.y, boxSize.x, boxSize.y);
            var strokeGeom = CreateGeometry(mako, strokeBox, type, sides, angle, lines);
            page.appendChild(IDOMPathNode.createStroked(mako, strokeGeom, brush));

            // Now draw filled version below it
            var fillBox = new FRect(strokeBox.x, strokeBox.y + boxSize.y * 1.1, boxSize.x, boxSize.y);
            var fillGeom = CreateGeometry(mako, fillBox, type, sides, angle, lines + 4);
            page.appendChild(IDOMPathNode.createFilled(mako, fillGeom, brush));

            // Move to next column and expand width 
            start.x += boxSize.x * 1.2;
            boxSize.x *= 1.3;

            lines += 1;
        }
    }

    private static IDOMPathGeometry CreateGeometry(IJawsMako mako, FRect box, ShapeType type, int sides, double angle, int lines)
    {
        return type switch
        {
            ShapeType.Ellipse => IDOMPathGeometry.createEllipse(mako, box),
            ShapeType.Hexagon => IDOMPathGeometry.createPolygon(mako, box, 6, angle),
            ShapeType.Polygon => IDOMPathGeometry.createPolygon(mako, box, (uint)sides, angle),
            ShapeType.Target => CreateTarget(mako, box, lines, angle),
            _ => IDOMPathGeometry.create(mako, box)
        };
    }

    private static IDOMPathGeometry CreateTarget(IJawsMako mako, FRect box, int nLines, double rotation)
    {
        var angle = rotation * Math.PI / 180.0;
        var incr = 2.0 * Math.PI / nLines;
        var radiusX = box.dX / 2.0;
        var radiusY = box.dY / 2.0;
        var centerX = box.x + radiusX;
        var centerY = box.y + radiusY;

        var builder = IDOMPathGeometryBuilder.create(mako);
        var center = new FPoint(centerX, centerY);
        builder.moveTo(center);

        for (var i = 0; i < nLines; i++)
        {
            var endX = centerX + radiusX * Math.Cos(angle);
            var endY = centerY + radiusY * Math.Sin(angle);
            builder.lineTo(new FPoint(endX, endY));
            builder.moveTo(center);
            angle += incr;
        }
        builder.close();
        return builder.createGeometry(mako, IDOMPathGeometry.eFillRule.eFRNonZero);
    }

    private static IDOMColor MakeDeviceNColor(IJawsMako mako, string name, CEDLVectDouble representation, IDOMColorSpace alternate)
    {
        // Create a vector of colorants with one entry
        var colorants = new CEDLVectColorantInfo();
        var colorant = new IDOMColorSpaceDeviceN.CColorantInfo(name, representation);
        colorants.append(colorant);

        // Create the DeviceN space
        var colorSpace = IDOMColorSpaceDeviceN.create(mako, colorants, alternate);

        // Return the new spot color
        return IDOMColor.create(mako, colorSpace, 1.0, 1.0);
    }
}

File: Visual Studio Projects/CreateColorsAndShapes/CreateColorsAndShapes/CreateColorsAndShapes.cpp
/* -----------------------------------------------------------------------
 * <copyright file="Main.cpp" company="Hybrid Software Helix Ltd">
 *  Copyright (c) 2025 Hybrid Software Helix Ltd. All rights reserved.
 * </copyright>
 * <summary>
 *  This example is provided on an "as is" basis and without warranty of any kind.
 *  Hybrid Software Helix Ltd. does not warrant or make any representations
 *  regarding the use or results of use of this example.
 * </summary>
 * -----------------------------------------------------------------------
 */

#include <iostream>
#include <jawsmako/jawsmako.h>
#include <jawsmako/pdfoutput.h>
#include <edl/idomcolorspace.h>
#include <jawsmako/layout.h>

using namespace JawsMako;
using namespace EDL;

IDOMPathGeometryPtr createHexagon(const IJawsMakoPtr& mako, FRect box, double rotation = 0);
IDOMPathGeometryPtr createTarget(const IJawsMakoPtr& mako, FRect box, double nLines, double rotation);
IDOMColorPtr MakeDeviceNColor(const IJawsMakoPtr& mako, const U8String& name, const CDoubleVect& representation, IDOMColorSpacePtr alternateSpace);

int main()
{
    try
    {
        const auto mako = IJawsMako::create();
        mako->enableAllFeatures(mako);
        const auto assembly = IDocumentAssembly::create(mako);
        const auto document = IDocument::create(mako);
        assembly->appendDocument(document);
        const auto page = IPage::create(mako);
        document->appendPage(page);
        const auto fixedPage = IDOMFixedPage::create(mako);
        page->setContent(fixedPage);

        // Create some process color brushes
        const auto cyan = IDOMSolidColorBrush::createSolidCmyk(mako, 1.0, 0.0, 0.0, 0.0);
        const auto magenta = IDOMSolidColorBrush::createSolidCmyk(mako, 0.0, 1.0, 0.0, 0.0);
        const auto yellow = IDOMSolidColorBrush::createSolidCmyk(mako, 0.0, 0.0, 1.0, 0.0);
        const auto black = IDOMSolidColorBrush::createSolidCmyk(mako, 0.0, 0.0, 0.0, 1.0);

        // Create some spot color brushes with types LAB, ICCBased and DeviceN

        // Create an LAB colorspace with D65 white point and -128 to 127 ranges for a and b values
        const auto labColorSpace = IDOMColorSpaceLAB::create(mako, 0.9504f, 1.0f, 1.0888f, 0.0, 0.0, 0.0, -128, 127, -128, 127);

        // Create a device CMYK colorspace from an ICC profile
        const auto iccBasedColorSpace = IDOMColorSpaceICCBased::create(
             mako, IDOMICCProfile::create(
                mako, IInputStream::createFromFile(
                    mako, R"(C:\Windows\System32\spool\drivers\color\WebCoatedFOGRA28.icc)")));
        
        // Create an LAB color
        const auto pantoneBlue072C_lab = IDOMColor::create(mako, labColorSpace, 1.0, 17.64, 43.0, -76.0);

        // Make a copy and convert to ICC space
        auto pantoneBlue072C_fogra = clone(pantoneBlue072C_lab, mako);
        pantoneBlue072C_fogra->setColorSpace(iccBasedColorSpace, eRelativeColorimetric, eBPCDefault, mako);

        // Create a DeviceN color
        const auto pantoneBlue072C_spot = MakeDeviceNColor(mako, "PANTONE BLUE 072 C", CDoubleVect{ 17.64, 43.0, -76.0 }, labColorSpace);

        // Create an All spot color
        const auto allColor = MakeDeviceNColor(mako, "All", CDoubleVect{ 1.0, 1.0, 1.0, 1.0 }, IDOMColorSpaceDeviceCMYK::create(mako));

        // Create spot color brushes
        const auto labBrush = IDOMSolidColorBrush::create(mako, pantoneBlue072C_lab);
        const auto iccBrush = IDOMSolidColorBrush::create(mako, pantoneBlue072C_fogra);
        const auto deviceNBrush = IDOMSolidColorBrush::create(mako, pantoneBlue072C_spot);
        const auto allBrush = IDOMSolidColorBrush::create(mako, allColor);

        // Draw some boxes
        const auto masterBoxSize = FPoint(90, 90);
        auto boxSize = masterBoxSize;
        auto origin = FPoint(85, 85);
        auto start = origin;

        // Draw Cyan box
        auto box = FRect(start.x, start.y, boxSize.x, boxSize.y);
        fixedPage->appendChild(IDOMPathNode::createFilled(mako, IDOMPathGeometry::create(mako, box), cyan));
        start.x += boxSize.x;

        // Draw Magenta box
        box = FRect(start.x, start.y, boxSize.x, boxSize.y);
        fixedPage->appendChild(IDOMPathNode::createFilled(mako, IDOMPathGeometry::create(mako, box), magenta));
        start.x += boxSize.x;

        // Draw Yellow box
        box = FRect(start.x, start.y, boxSize.x, boxSize.y);
        fixedPage->appendChild(IDOMPathNode::createFilled(mako, IDOMPathGeometry::create(mako, box), yellow));
        start.x += boxSize.x;

        // Draw Black box
        box = FRect(start.x, start.y, boxSize.x, boxSize.y);
        fixedPage->appendChild(IDOMPathNode::createFilled(mako, IDOMPathGeometry::create(mako, box), black));
        start.x += boxSize.x;

        // Draw LAB box 
        box = FRect(start.x, start.y, boxSize.x, boxSize.y);
        fixedPage->appendChild(IDOMPathNode::createFilled(mako, IDOMPathGeometry::create(mako, box), labBrush));
        start.x += boxSize.x;

        // Draw ICC box
        box = FRect(start.x, start.y, boxSize.x, boxSize.y);
        fixedPage->appendChild(IDOMPathNode::createFilled(mako, IDOMPathGeometry::create(mako, box), iccBrush));
        start.x += boxSize.x;

        // Draw DeviceN box
        box = FRect(start.x, start.y, boxSize.x, boxSize.y);
        fixedPage->appendChild(IDOMPathNode::createFilled(mako, IDOMPathGeometry::create(mako, box), deviceNBrush));
        start.x += boxSize.x;
        
        // Draw a border in All
        const uint32 strokeWidth = 20;
        box = FRect(origin.x - strokeWidth / 2.0, origin.y - strokeWidth / 2.0, (boxSize.x * 7) + strokeWidth, boxSize.y + strokeWidth);
        fixedPage->appendChild(IDOMPathNode::createStroked(mako, IDOMPathGeometry::create(mako, box), allBrush, FMatrix(),
            IDOMPathGeometryPtr(), strokeWidth));

        // Draw some other shapes below
        boxSize = masterBoxSize;
        origin.y += 180;
        start = origin;

        // Draw Cyan circle
        box = FRect(start.x, start.y, boxSize.x, boxSize.y);
        fixedPage->appendChild(IDOMPathNode::createStroked(mako, IDOMPathGeometry::createEllipse(mako, box), cyan));
        box.y += boxSize.y * 1.1;
        fixedPage->appendChild(IDOMPathNode::createFilled(mako, IDOMPathGeometry::createEllipse(mako, box), cyan));

        // Draw Magenta circle
        start.x += boxSize.x * 1.2;
        boxSize.x *= 1.3;
        box = FRect(start.x, start.y, boxSize.x, boxSize.y);
        fixedPage->appendChild(IDOMPathNode::createStroked(mako, IDOMPathGeometry::createEllipse(mako, box), magenta));
        box.y += boxSize.y * 1.1;
        fixedPage->appendChild(IDOMPathNode::createFilled(mako, IDOMPathGeometry::createEllipse(mako, box), magenta));

        // Draw Yellow circle
        start.x += boxSize.x * 1.2;
        boxSize.x *= 1.3;
        box = FRect(start.x, start.y, boxSize.x, boxSize.y);
        fixedPage->appendChild(IDOMPathNode::createStroked(mako, IDOMPathGeometry::createEllipse(mako, box), yellow));
        box.y += boxSize.y * 1.1;
        fixedPage->appendChild(IDOMPathNode::createFilled(mako, IDOMPathGeometry::createEllipse(mako, box), yellow));

        // Draw Black circle
        start.x += boxSize.x * 1.2;
        boxSize.x *= 1.3;
        box = FRect(start.x, start.y, boxSize.x, boxSize.y);
        fixedPage->appendChild(IDOMPathNode::createStroked(mako, IDOMPathGeometry::createEllipse(mako, box), black));
        box.y += boxSize.y * 1.1;
        fixedPage->appendChild(IDOMPathNode::createFilled(mako, IDOMPathGeometry::createEllipse(mako, box), black));

        // Draw some hexagons
        boxSize = masterBoxSize;
        origin.y += 200;
        start = origin;

        // Draw LAB hexagon
        box = FRect(start.x, start.y, boxSize.x, boxSize.y);
        fixedPage->appendChild(IDOMPathNode::createStroked(mako, createHexagon(mako, box), labBrush));
        box.y += boxSize.y * 1.1;
        fixedPage->appendChild(IDOMPathNode::createFilled(mako, createHexagon(mako, box), labBrush));

        // Draw ICC hexagon
        start.x += boxSize.x * 1.2;
        boxSize.x *= 1.3;
        box = FRect(start.x, start.y, boxSize.x, boxSize.y);
        fixedPage->appendChild(IDOMPathNode::createStroked(mako, createHexagon(mako, box), iccBrush));
        box.y += boxSize.y * 1.1;
        fixedPage->appendChild(IDOMPathNode::createFilled(mako, createHexagon(mako, box), iccBrush));

        // Draw DeviceN hexagon
        start.x += boxSize.x * 1.2;
        boxSize.x *= 1.3;
        box = FRect(start.x, start.y, boxSize.x, boxSize.y);
        fixedPage->appendChild(IDOMPathNode::createStroked(mako, createHexagon(mako, box), deviceNBrush));
        box.y += boxSize.y * 1.1;
        fixedPage->appendChild(IDOMPathNode::createFilled(mako, createHexagon(mako, box), deviceNBrush));

        // Draw All color hexagon
        start.x += boxSize.x * 1.2;
        boxSize.x *= 1.3;
        box = FRect(start.x, start.y, boxSize.x, boxSize.y);
        fixedPage->appendChild(IDOMPathNode::createStroked(mako, createHexagon(mako, box), allBrush));
        box.y += boxSize.y * 1.1;
        fixedPage->appendChild(IDOMPathNode::createFilled(mako, createHexagon(mako, box), allBrush));

        // Draw some polygons
        auto sides = 8;
        auto angle = 22.5;
        boxSize = masterBoxSize;
        origin.y += 200;
        start = origin;

        // Draw Cyan polygon
        box = FRect(start.x, start.y, boxSize.x, boxSize.y);
        fixedPage->appendChild(IDOMPathNode::createStroked(mako, IDOMPathGeometry::createPolygon(mako, box, sides, angle), cyan));
        box.y += boxSize.y * 1.1;
        fixedPage->appendChild(IDOMPathNode::createFilled(mako, IDOMPathGeometry::createPolygon(mako, box, sides, angle), cyan));

        // Draw Magenta polygon
        start.x += boxSize.x * 1.2;
        boxSize.x *= 1.3;
        box = FRect(start.x, start.y, boxSize.x, boxSize.y);
        fixedPage->appendChild(IDOMPathNode::createStroked(mako, IDOMPathGeometry::createPolygon(mako, box, sides, angle), magenta));
        box.y += boxSize.y * 1.1;
        fixedPage->appendChild(IDOMPathNode::createFilled(mako, IDOMPathGeometry::createPolygon(mako, box, sides, angle), magenta));

        // Draw Yellow polygon
        start.x += boxSize.x * 1.2;
        boxSize.x *= 1.3;
        box = FRect(start.x, start.y, boxSize.x, boxSize.y);
        fixedPage->appendChild(IDOMPathNode::createStroked(mako, IDOMPathGeometry::createPolygon(mako, box, sides, angle), yellow));
        box.y += boxSize.y * 1.1;
        fixedPage->appendChild(IDOMPathNode::createFilled(mako, IDOMPathGeometry::createPolygon(mako, box, sides, angle), yellow));

        // Draw Black polygon
        start.x += boxSize.x * 1.2;
        boxSize.x *= 1.3;
        box = FRect(start.x, start.y, boxSize.x, boxSize.y);
        fixedPage->appendChild(IDOMPathNode::createStroked(mako, IDOMPathGeometry::createPolygon(mako, box, sides, angle), black));
        box.y += boxSize.y * 1.1;
        fixedPage->appendChild(IDOMPathNode::createFilled(mako, IDOMPathGeometry::createPolygon(mako, box, sides, angle), black));

        // ----- Draw some targets -----
        
        auto lines = 3;
        angle = 0.0;
        boxSize = masterBoxSize;
        origin.y += 200;
        start = origin;

        // Draw LAB polygon
        box = FRect(start.x, start.y, boxSize.x, boxSize.y);
        fixedPage->appendChild(IDOMPathNode::createStroked(mako, createTarget(mako, box, lines, angle), labBrush));
        box.y += boxSize.y * 1.1;
        fixedPage->appendChild(IDOMPathNode::createStroked(mako, createTarget(mako, box, lines + 4, angle), labBrush));

        // Draw ICC polygon
        start.x += boxSize.x * 1.2;
        boxSize.x *= 1.3;
        box = FRect(start.x, start.y, boxSize.x, boxSize.y);
        lines += 1;
        fixedPage->appendChild(IDOMPathNode::createStroked(mako, createTarget(mako, box, lines, angle), iccBrush));
        box.y += boxSize.y * 1.1;
        fixedPage->appendChild(IDOMPathNode::createStroked(mako, createTarget(mako, box, lines + 4, angle), iccBrush));

        // Draw DeviceN polygon
        start.x += boxSize.x * 1.2;
        boxSize.x *= 1.3;
        box = FRect(start.x, start.y, boxSize.x, boxSize.y);
        lines += 1;
        fixedPage->appendChild(IDOMPathNode::createStroked(mako, createTarget(mako, box, lines, angle), deviceNBrush));
        box.y += boxSize.y * 1.1;
        fixedPage->appendChild(IDOMPathNode::createStroked(mako, createTarget(mako, box, lines + 4, angle), deviceNBrush));

        // Draw All color polygon
        start.x += boxSize.x * 1.2;
        boxSize.x *= 1.3;
        box = FRect(start.x, start.y, boxSize.x, boxSize.y);
        lines += 1;
        fixedPage->appendChild(IDOMPathNode::createStroked(mako, createTarget(mako, box, lines, angle), allBrush));
        box.y += boxSize.y * 1.1;
        fixedPage->appendChild(IDOMPathNode::createStroked(mako, createTarget(mako, box, lines + 4, angle), allBrush));

        IPDFOutput::create(mako)->writeAssembly(assembly, "test.pdf");
    }
    catch (IError& e)
    {
        const String errorFormatString = getEDLErrorString(e.getErrorCode());
        std::wcerr << L"Exception thrown: " << e.getErrorDescription(errorFormatString) << std::endl;
        return static_cast<int>(e.getErrorCode());
    }
    catch (std::exception& e)
    {
        std::wcerr << L"std::exception thrown: " << e.what() << std::endl;
        return 1;
    }

    return 0;
}

IDOMPathGeometryPtr createHexagon(const IJawsMakoPtr& mako, FRect box, double rotation)
{
    return IDOMPathGeometry::createPolygon(mako, box, 6, rotation);
}

IDOMPathGeometryPtr createTarget(const IJawsMakoPtr& mako, FRect box, double nLines, double rotation)
{
    double angle = rotation * (PI / 180.0);
    const double incr = 2.0 * PI / nLines;

    const auto radiusX = box.dX / 2.0;
    const auto radiusY = box.dY / 2.0;
    const auto centerX = box.x + radiusX;
    const auto centerY = box.y + radiusY;

    const auto geometryBuilder = IDOMPathGeometryBuilder::create(mako);
    const auto center = FPoint(centerX, centerY);
    geometryBuilder->moveTo(center);

    double endX = 0.0, endY = 0.0;
    for (int i = 0; i < nLines; i++)
    {
        endX = centerX + radiusX * cos(angle);
        endY = centerY + radiusY * sin(angle);
        geometryBuilder->lineTo(FPoint(endX, endY));
        geometryBuilder->moveTo(center);
        angle += incr;
    }
    geometryBuilder->close();
    return geometryBuilder->createGeometry(mako, IDOMPathGeometry::eFRNonZero);
}

IDOMColorPtr MakeDeviceNColor(const IJawsMakoPtr& mako, const U8String& name, const CDoubleVect& representation, IDOMColorSpacePtr alternateSpace)
{
    // Create a vector of colorants with one entry
    auto colorants = IDOMColorSpaceDeviceN::CColorantInfoVect();
    const auto colorantInfo = IDOMColorSpaceDeviceN::CColorantInfo(name, representation);
    colorants.append(colorantInfo);

    // Create the DeviceN space
    const auto colorSpace = IDOMColorSpaceDeviceN::create(mako, colorants, alternateSpace);

    // Return the new spot color
    return IDOMColor::create(mako, colorSpace, 1.0, 1.0);
}